File: medium/649. Dota2 Senate/649. Dota2 Senate.py
from collections import deque


class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        # t = {"R": "Radiant", "D": "Dire"}

        # if len(senate) == 1:
        #     return t[senate[0]]

        # queue = list(senate)

        # while queue != 1:
        #     curr = queue.pop(0)

        #     for i in range(len(queue)):
        #         if curr != queue[i]:
        #             queue.pop(i)
        #             queue.append(curr)
        #             break

        #     if not queue:
        #         return t[curr]

        # return t[queue[0]]

        r_queue = deque()
        d_queue = deque()
        n = len(senate)

        for i, s in enumerate(senate):
            if s == "R":
                r_queue.append(i)
            else:
                d_queue.append(i)

        while d_queue and r_queue:
            r = r_queue.popleft()
            d = d_queue.popleft()

            if r < d:
                r_queue.append(r + n)
            else:
                d_queue.append(d + n)

        return "Radiant" if r_queue else "Dire"


solution = Solution()
print(solution.predictPartyVictory(senate="RD"))
print(solution.predictPartyVictory(senate="RDD"))
print(solution.predictPartyVictory(senate="R"))
print(
    solution.predictPartyVictory(
        senate="RRDRDRDRDRDRRRDRRDRDRDRDRDRDRDRDRRDRDDDDDDDRDRRDRDRDRRD"
    )
)
print(solution.predictPartyVictory(senate="DRRDRDRDRDDRDRDR"))
print(solution.predictPartyVictory(senate="DDRRR"))

File: medium/649. Dota2 Senate/649. Dota2 Senate.go
package main

func predictPartyVictory(senate string) string {
	rItems := []int{}
	dItems := []int{}
	n := len(senate)

	for i := 0; i < n; i++ {
		if senate[i] == 'R' {
			rItems = append(rItems, i)
		} else {
			dItems = append(dItems, i)
		}
	}

	for len(rItems) > 0 && len(dItems) > 0 {
		r := rItems[0]
		rItems = rItems[1:]

		d := dItems[0]
		dItems = dItems[1:]

		if r < d {
			rItems = append(rItems, n+r)
		} else {
			dItems = append(dItems, n+d)
		}
	}

	if len(rItems) > 0 {
		return "Radiant"
	} else {
		return "Dire"
	}
}
